source/core/TypeAndDataPathFinder.cpp

```diff
@@ -0,0 +1,241 @@
+    virtual void VisitEnum(EnumType* type);
+    virtual void VisitRefNumVal(RefNumValType* type);
+    virtual void VisitDefaultValue(DefaultValueType* type);
+    virtual void VisitDefaultPointer(DefaultPointerType* type);
+    virtual void VisitCustomDataProc(CustomDataProcType* type);
+};
+//------------------------------------------------------------
+void TypeManager::GetPathFromPointer(TypeRef tNeedle, DataPointer pNeedle, StringRef path)
+{
+    path->Resize1D(0);
+    PathFinderVisitor drv(tNeedle, pNeedle, path);
+    for (TypeManager *tm = this; tm && !drv.Found(); tm = tm->BaseTypeManager()) {
+        drv.Accept(tm);
+    }
+
+    if (!drv.Found()) {
+        path->AppendCStr("*pointer-not-found*");
+    }
+}
+//------------------------------------------------------------
+PathFinderVisitor::PathFinderVisitor(TypeRef tHaystack, DataPointer pNeedle, StringRef path)
+{
+    _tNeedle = tHaystack;
+    _pNeedle = pNeedle;
+    _pHayStack = null;
+    _pathEmpty = true;
+    _found = false;
+    _path = path;
+}
+//------------------------------------------------------------
+//! Visitor dispatch helper method that traverses all the types in a TypeManager
+void PathFinderVisitor::Accept(TypeManagerRef tm)
+{
+    TypeRef type = tm->TypeList();
+    while (type) {
+        Accept(type, type->Begin(kPASoftRead));
+        if (_found)
+            break;
+        type = type->Next();
+    }
+
+    SubString rootName;
+    Boolean isTypeRefConstant = false;
+    if (_found) {
+        rootName = type->Name();
+    } else {
+        _found = tm->PointerToTypeConstRefName((TypeRef*)_pNeedle, &rootName);
+        isTypeRefConstant = _found;
+    }
+
+    if (_found) {
+        if (isTypeRefConstant) {
+            _path->InsertCStr(0, ".");
+        }
+        if (!type->IsA("VirtualInstrument")) {
+            _path->InsertSubString(0, &rootName);
+        }
+    }
+}
+//------------------------------------------------------------
+//! Visitor dispatch helper method that defines a new/smaller haystack
+void PathFinderVisitor::Accept(TypeRef tHayStack, DataPointer pHayStack)
+{
+    if (_pNeedle == tHayStack) {
+        // The value may be a TypeRef. If so check and catch it here,
+        // no need for each visitor to do a check.
+        // _isTypeRef = true;
+        _found = true;
+    } else if ((_pNeedle == pHayStack) && (_tNeedle->IsA(tHayStack, true) )) {
+        // Once the needle is found then there is no need to keep
+        // digging if the types match. However, if the needle is the
+        // first field in a cluster more digging may be necessary
+        // to get the correct full symbol path.
+        _found = true;
+    } else {
+        // Dig deeper into the data.
+        void* saveData = _pHayStack;
+        _pHayStack = pHayStack;
+        tHayStack->Accept(this);
+        _pHayStack = saveData;
+    }
+}
+//------------------------------------------------------------
+void PathFinderVisitor::VisitBad(TypeRef type)
+{
```

Comment:
Can you move the empty Visit members and the trivial ones that just call Accept() to be inline in the class definition?